core/invoker/src/main/scala/whisk/core/invoker/InvokerReactive.scala

```diff
@@ -130,30 +129,23 @@ class InvokerReactive(
             s"posted ${if (recovery) "recovery" else "completion"} of activation ${activationResult.activationId}")
       }
     }
-    // Potentially sends activation metadata to kafka if user events are enabled
-    UserEvents.send(
-      producer, {
-        val activation = Activation(
-          activationResult.namespace + EntityPath.PATHSEP + activationResult.name,
-          activationResult.response.statusCode,
-          activationResult.duration.getOrElse(0),
-          activationResult.annotations.getAs[Long](WhiskActivation.waitTimeAnnotation).getOrElse(0),
-          activationResult.annotations.getAs[Long](WhiskActivation.initTimeAnnotation).getOrElse(0),
-          activationResult.annotations.getAs[String](WhiskActivation.kindAnnotation).getOrElse("unknown_kind"),
-          activationResult.annotations.getAs[Boolean](WhiskActivation.conductorAnnotation).getOrElse(false),
-          activationResult.annotations
-            .getAs[ActionLimits](WhiskActivation.limitsAnnotation)
-            .map(al => al.memory.megabytes)
-            .getOrElse(0),
-          activationResult.annotations.getAs[Boolean](WhiskActivation.causedByAnnotation).getOrElse(false))
+
+    if (UserEvents.enabled) {
+      val event = Activation.from(activationResult).map { body =>
         EventMessage(
           s"invoker${instance.instance}",
-          activation,
+          body,
           activationResult.subject,
           activationResult.namespace.toString,
           userId,
-          activation.typeName)
-      })
+          body.typeName)
+      }
+
+      event match {
+        case Success(msg) => UserEvents.send(producer, msg)
+        case Failure(t)   => logging.warn(this, s"activation event was not sent: $t")
```

Comment:
Action fqn and kind are always defined, if it is not the case, should the level be error then?

Reply:
Your wish has been granted.